### PR TITLE
3918 Migrate V3 Opinion Search API to ES

### PR DESCRIPTION
**Pull request by albertisfu**
According to #3918, this PR has migrated the V3 Opinions Search API to utilize ES as its search engine.

The migration is controlled by the flag **`o-es-search-api-active`**, allowing for independent management from the frontend search.

The API uses the same serializer as the Opinion Search Alerts webhook. So the changes we need to inform users are:

**Removed fields:**
- **`caseNameShort`**
- **`pagerank`**
- **`status_exact`**
- **`non_participating_judges`**
- **`source`**

**New field:**
- **`date_created`**

Additionally, the content of the **`snippet`** has changed. In the Solr version, it included content from all fields, while now it displays only the Opinion text content.

### **API Results Count**

- The API displays Opinions as results collapsed by **`cluster_id`**. However, the total number of hits returned by a [collapse query](https://www.elastic.co/guide/en/elasticsearch/reference/current/collapse-search-results.html#collapse-search-results) reflects the matched results without collapsing. This is important due to this count is used for pagination.
- To resolve the count issue, a cardinality aggregation by **`cluster_id`** is performed, balancing memory usage and accuracy. This aggregation includes a **`precision_threshold`** setting that defines a threshold below which the count will be accurate. Beyond this threshold, the count may have an error of 1-6%. I've set the threshold to the maximum of **`40,000`**, meaning that if the number of collapsed results exceeds this, the actual count could vary. This might impact users doing deep pagination to access more than **`40,000`** results, potentially leading to extra blank pages or missing pages. If this becomes a concern, we could switch to a different, more resource-intensive but exact aggregation method.

- With this change, we're no longer using Solr in the selenium tests, so I've removed all Solr-related code from `BaseSeleniumTest`.
- This PR is dependent on #3911, so it should either be merged first or incorporated into it.

### CODE REVIEW

**Comment by mlissner** [comment]
I'll hold off on reviewing this until #3918 is resolved/merged, but I did want to respond to this:

> This might impact users doing deep pagination to access more than 40,000 

Isn't this API blocked after page 100 or so? If so, should we make the cardinality aggregation accurate to whatever result number that would imply? E.g., `x results/page × y pages = x×y` items?

**Comment by albertisfu** [comment]
> Isn't this API blocked after page 100 or so? If so, should we make the cardinality aggregation accurate to whatever result number that would imply? E.g., x results/page × y pages = x×y items?

It doesn't seem like deep pagination is blocked in the Search API. For instance, here is a query involving page 1000:
https://www.courtlistener.com/api/rest/v3/search/?page=1000&q=California

Should we add a limit after page 100 for all the V3 Search API endpoints?

**Comment by mlissner** [comment]
I'm amazed that performs. In v4, let's cut off the search API after page 100, and adjust the cardinality accordingly. 

For switching to the ES-based v3, let's just tell people that alert counts are accurate up to 2000 items (page 100, right?), and are approximate after that point. That seems like an entirely fine compromise.

**Comment by albertisfu** [comment]
Got it for V4 I'll consider using `ShallowOnlyPageNumberPagination`, which has a `max_pagination_depth` of 100.

The reason the V3 Search API is not limited to 100 pages is that it uses the default paginator: `PageNumberPagination`.

- For now, in V3, I have changed the `precision_threshold` in the aggregation count to 2000. So, if the returned results are more than this number, the count of results and the pagination (after page 100) might not be accurate. This impacts only the V3 Search API.
- Alerts are not affected since, by default, only up to 20 items are included per alert. However, I noticed an issue related to this count in the Oral Arguments Scheduled alerts. Since we only store up to 30 hits per alert, we only show `30+` results in the alert. To align with that behavior in the Opinion Search Alerts handled by `cl_send_alerts`, I tweaked it to show the `+` if the number of results equals the `hits_limit` (20).
- Added a Pagination test for the Opinion Search API.

- Finally, this ES version of the V3 Opinion Search API is behind the waffle: `o-es-search-api-active`. I tested the waffle settings due to last time the waffle was created as enabled by default. So, I found that the correct way to set them in K8s is:
  `WAFFLE_FLAG_DEFAULT=off`
  `WAFFLE_SWITCH_DEFAULT=off`
  Instead of `"off"`

**Comment by mlissner** [comment]
> only up to 20 items are included per alert

Would it be simpler if all alert types just had 20 items? It sounds like oral arguments has 30 for no particular reason? 

> So, I found that the correct way to set them in K8s is...

Hm, we set these in a yaml file for k8s, which has them as: 

```
  WAFFLE_FLAG_DEFAULT: "off"
  WAFFLE_SWITCH_DEFAULT: "off"
```

I can't remove the quotes without breaking the yaml format. Hm. This shouldn't be hard. I thought that django-environ accepted practically anything as a boolean value.

**Review by mlissner — COMMENTED** [review]
This looks good to me. Just a couple thoughts, then I think off to Eduardo for his more thorough review.

**Comment by albertisfu** [comment]
> Would it be simpler if all alert types just had 20 items? It sounds like oral arguments has 30 for no particular reason?

Yeah in OA we decided to go a but higher to 30. But modifying this number is easy. I have now updated the number of scheduled alerts in Oral Arguments to 20. This is controlled by the setting `SCHEDULED_ALERT_HITS_LIMIT`. Nonetheless, this adjustment doesn't eliminate the requirement to display `"20+ hits"` because, when using percolator for scheduled alerts, the exact number of alerts triggered is unknown. We only store alerts up to the `SCHEDULED_ALERT_HITS_LIMIT`.

The only side effect of this modification is that, upon merging, if there are more than 20 alerts OA scheduled at that moment for some users, they will receive all the scheduled alerts when the command is executed. However, the email will indicate `"20+ hits."` even if they see more in the email, next day this will normalize showing only up to 20 alerts in the email.

> I can't remove the quotes without breaking the yaml format. Hm. This shouldn't be hard. I thought that django-environ accepted practically anything as a boolean value.

You're correct. I repeated the experiment, and "off" was properly recognized after I restarted  docker-compose.
I'm now unsure why the waffle was initially created as activated. To prevent this from happening again I've better created the [o-es-search-api-active](https://www.courtlistener.com/admin/waffle/flag/10/change/) flag in the admin. So it won't be automatically created upon merge.

File: cl/search/api_utils.py
Comment (ERosendo) [inline cl/search/api_utils.py:46]: Wouldn't it be more consistent to name this variable `is_opinion_active`? The other flags related to ES don't include "es" in their names.

File: cl/alerts/management/commands/cl_send_alerts.py
Comment (ERosendo) [inline cl/alerts/management/commands/cl_send_alerts.py:36]: Can we use the `SCHEDULED_ALERT_HITS_LIMIT` setting to avoid hardcoding this value?

**Comment by ERosendo** [comment]
The code looks good 👍 ! Let's address the two comments and then merge.

File: cl/search/api_utils.py
Comment (albertisfu) [reply cl/search/api_utils.py]: Sure, I've updated the variable name to `is_opinion_active`

File: cl/alerts/management/commands/cl_send_alerts.py
Comment (albertisfu) [reply cl/alerts/management/commands/cl_send_alerts.py]: Yeah, now `SCHEDULED_ALERT_HITS_LIMIT ` is also equals to `20`, changed it.

**Comment by albertisfu** [comment]
Thanks @ERosendo I've applied your suggestions! 

The new [o-es-search-api-active](https://www.courtlistener.com/admin/waffle/flag/10/change/) flag has been created as disabled, so we can merge this one safely.

**Review by ERosendo — APPROVED** [review]

**Comment by mlissner** [comment]
I'm happy if you guys are. Set for auto-merge. Thank you both!

**Pull request merged**
